File: src/engine/random/prng.ts
/**
 * UI Explorer — Seeded Pseudo-Random Number Generator (PRNG)
 * Implements deterministic Mulberry32 algorithm.
 * Guarantees identical output for identical seeds across sessions.
 */

export interface PRNG {
  next(): number; // 0..1
  nextInt(min: number, max: number): number; // min..max inclusive
  nextFloat(min: number, max: number): number;
  choice<T>(array: T[]): T;
  weightedChoice<T>(options: { value: T; weight: number }[]): T;
}

/**
 * Converts any string or numeric seed into a 32-bit positive integer hash.
 */
export function hashSeed(seed: string | number): number {
  if (typeof seed === 'number') {
    return Math.floor(Math.abs(seed)) || 1;
  }
  let hash = 0;
  for (let i = 0; i < seed.length; i++) {
    const char = seed.charCodeAt(i);
    hash = (hash << 5) - hash + char;
    hash |= 0; // Convert to 32bit integer
  }
  return Math.abs(hash) || 12345;
}

/**
 * Creates a deterministic Mulberry32 PRNG instance.
 */
export function createSeededRandom(seedInput: string | number): PRNG {
  let s = hashSeed(seedInput);

  const next = (): number => {
    let t = (s += 0x6d2b79f5);
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };

  const nextInt = (min: number, max: number): number => {
    return Math.floor(next() * (max - min + 1)) + min;
  };

  const nextFloat = (min: number, max: number): number => {
    return next() * (max - min) + min;
  };

  const choice = <T>(array: T[]): T => {
    if (!array || array.length === 0) {
      throw new Error('PRNG.choice called on empty array');
    }
    return array[Math.floor(next() * array.length)];
  };

  const weightedChoice = <T>(options: { value: T; weight: number }[]): T => {
    if (!options || options.length === 0) {
      throw new Error('PRNG.weightedChoice called on empty options');
    }
    const totalWeight = options.reduce((sum, item) => sum + Math.max(0, item.weight), 0);
    if (totalWeight <= 0) {
      return options[0].value;
    }
    let randomVal = next() * totalWeight;
    for (const item of options) {
      const w = Math.max(0, item.weight);
      if (randomVal < w) {
        return item.value;
      }
      randomVal -= w;
    }
    return options[options.length - 1].value;
  };

  return {
    next,
    nextInt,
    nextFloat,
    choice,
    weightedChoice
  };
}

File: tools/style-engine/style_engine/prng.py
"""
Seeded PRNG — Mulberry32, bit-for-bit compatible with src/engine/random/prng.ts.

Every draw the browser engine makes has a counterpart here, in the same order,
so the same seed yields the same style on both sides.
"""

import math

MASK32 = 0xFFFFFFFF


def _to_int32(value):
    """JavaScript ToInt32."""
    value &= MASK32
    return value - (1 << 32) if value & 0x80000000 else value


def hash_seed(seed):
    """Turn any string or number into the positive 32-bit integer the TS engine uses."""
    if isinstance(seed, bool):
        seed = int(seed)
    if isinstance(seed, (int, float)):
        return int(math.floor(abs(seed))) or 1
    h = 0
    for ch in str(seed):
        h = _to_int32((h << 5) - h + ord(ch))
    return abs(h) or 12345


class SeededRandom:
    """Deterministic Mulberry32 generator."""

    def __init__(self, seed):
        self.state = hash_seed(seed)

    def next(self):
        self.state = (self.state + 0x6D2B79F5) & MASK32
        t = self.state
        t = ((t ^ (t >> 15)) * (t | 1)) & MASK32
        t ^= (t + (((t ^ (t >> 7)) * (t | 61)) & MASK32)) & MASK32
        t &= MASK32
        return ((t ^ (t >> 14)) & MASK32) / 4294967296.0

    def next_int(self, lo, hi):
        return int(math.floor(self.next() * (hi - lo + 1))) + lo

    def next_float(self, lo, hi):
        return self.next() * (hi - lo) + lo

    def choice(self, items):
        if not items:
            raise ValueError("PRNG.choice called on empty list")
        return items[int(math.floor(self.next() * len(items)))]

    def weighted_choice(self, options):
        """options: list of (value, weight) tuples."""
        if not options:
            raise ValueError("PRNG.weighted_choice called on empty options")
        total = sum(max(0, w) for _, w in options)
        if total <= 0:
            return options[0][0]
        r = self.next() * total
        for value, weight in options:
            w = max(0, weight)
            if r < w:
                return value
            r -= w
        return options[-1][0]


def js_round(x):
    """JavaScript Math.round (half towards +infinity)."""
    return math.floor(x + 0.5)


def js_num(x):
    """Format a float the way JavaScript's Number#toString does for the values we emit."""
    if isinstance(x, bool):
        return str(x).lower()
    if isinstance(x, int) or float(x).is_integer():
        return str(int(x))
    return repr(float(x))
